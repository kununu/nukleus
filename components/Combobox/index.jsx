// NOTE: we make extensive use of property initialiser for binding methods to the class instance:
// onChange, getSuggestionValue etc.

import React, {Component, PropTypes} from 'react';
import Autosuggest from 'react-autosuggest';

import styles from './index.scss';

import {
  controlLabel,
  errorStyles,
  formControl,
  formGroup
} from '../index.scss';

function getSuggestions (value, items = []) {
  const inputValue = value.trim().toLowerCase();
  const inputLength = inputValue.length;
  return items
    .filter(item => item && item.toLowerCase().slice(0, inputLength) === inputValue)
    .map(item => ({value: item}));
}

export default class ComboboxComponent extends Component {
  static propTypes = {
    disabled: PropTypes.bool,
    error: PropTypes.string,
    id: PropTypes.string.isRequired,
    inputStyles: PropTypes.string,
    inputValue: PropTypes.string,
    items: PropTypes.array,
    label: PropTypes.string,
    name: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
    required: PropTypes.bool
  };

  static defaultProps = {
    disabled: false,
    error: null,
    inputStyles: 'inline',
    inputValue: '',
    items: [],
    label: '',
    placeholder: '',
    required: false
  };

  state = {
    showError: false,
    suggestions: getSuggestions('', this.props.items),
    value: this.props.inputValue
  };

  componentWillReceiveProps (nextProps) {
    if (nextProps.error) this.showError();
    if (nextProps.inputValue !== this.props.inputValue) {
      this.setState({
        value: nextProps.inputValue
      });
    }
  }

  onSuggestionsFetchRequested = ({value}) => {
    this.setState({
      suggestions: getSuggestions(value, this.props.items)
    });
  };

  onChange = (event, {newValue}) => {
    this.hideError();
    this.setState({
      value: newValue
    });
  };

  getSuggestionValue = suggestion => suggestion.value;

  handleSelection = (e, {method}) => {
    if (method === 'enter') {
      e.preventDefault();
    }
  };

  showError () {
    this.setState({showError: true});
  }

  hideError () {
    this.setState({showError: false});
  }

  renderSuggestion = suggestion => <span>{suggestion.value}</span>;

  render () {
    const {
      name,
      id,
      label,
      error,
      required,
      inputStyles,
      placeholder,
      disabled
    } = this.props;

    return (
      <div className={`${formGroup} ${styles[inputStyles]}`}>
        <label
          className={controlLabel}
          htmlFor="city">

          {label}
        </label>

        <div className={styles.container}>
          <Autosuggest
            suggestions={this.state.suggestions}
            theme={styles}
            onSuggestionSelected={this.handleSelection}
            onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
            shouldRenderSuggestions={() => true}
            onSuggestionsClearRequested={() => true}
            getSuggestionValue={this.getSuggestionValue}
            renderSuggestion={this.renderSuggestion}
            inputProps={{
              className: formControl,
              disabled,
              id,
              name,
              onChange: this.onChange,
              placeholder,
              required,
              value: this.state.value
            }} />

          <span className={styles.handle}>
            <i className="fa fa-caret-down" aria-hidden="true" />
          </span>

          {this.state.showError &&
            <span className={errorStyles}>{error}</span>}
        </div>
      </div>
    );
  }
}
