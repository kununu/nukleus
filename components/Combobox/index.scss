@import '~kununu-theme/scss/base/variables';

.input {
  cursor: pointer;
}

.inline {
  display: flex;
  flex-flow: row;
  align-items: center;
  width: 100%;

  label {
    margin-right: 10px;
    margin-bottom: 0;
    width: 140px;

    @media (max-width: $screen-xs-min) {
      text-align: left;
    }
  }

  @media (max-width: $screen-xs-min) {
    display: block;
  }
}

.container {
  position: relative;
  width: 100%;
  line-height: 1.65;

  * {
    box-sizing: border-box;
  }

  // Inline error definition
  // Consider abstracting out to another component if we start using this a lot
  .error {
    border-radius: 4px;
    font-size: 11px;
    padding: 5px 12px;
    text-transform: uppercase;
    font-weight: $font-weight-regular;
    color: $white;
    position: absolute;
    z-index: 1;
    left: 0;
    top: 45px;

    &:after {
      content: '';
      position: absolute;
      display: block;
      top: -5px;
      right: 45%;
      bottom: auto;
      width: 0;
      height: 0;
      border-style: solid;
      border-width: 0 6px 6px;
      border-color: transparent transparent $brand-danger;
    }
  }
}

.itemsContainer {
  box-shadow: 0 5px 10px rgba($black, .2);
  border-radius: $border-radius-base;
  background-clip: padding-box;
  font-size: $font-size-base;
  min-width: 250px;
  width: 100%;
  margin-top: 2px;
  background-color: $white;
  max-height: 300px;
  overflow: auto;
  position: absolute;
  top: 100%;
  left: 0;
  z-index: 2;

  @media (max-width: $screen-xs-min) {
    left: 0;
    right: 0;
    width: 100%;
  }
}

.suggestionsList {
  list-style: none;
  padding: 0;
  margin: 0;
}

.suggestion {
  text-align: left;
  padding: 10px 20px;
}

.suggestionFocused {
  background: $blue;
  color: $white;
  cursor: pointer;
}

.smallText {
  color: $gray-base-40;
  font-style: italic;
}

.handle {
  position: absolute;
  pointer-events: none;
  right: 10px;
  top: 10px;

  i {
    color: $gray;
    font-size: $font-size-base;
  }
}
