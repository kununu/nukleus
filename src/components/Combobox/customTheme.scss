@import '~@kununu/kununu-theme-v2/scss/base/variables';

.suggestionsContainer {
  box-shadow: 0 2px 10px rgba($black, .2);
  font-size: 12px;
  width: 80%;

  @media (max-width: $screen-xs-max) {
    width: auto;
    border-radius: 0;
  }

  ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }
}

.formControl {
  width: 100%;
  min-height: 50px;
  padding: 5px;

  &:focus {
    border: 1px solid $green;
    box-shadow: inset 0 0 0 1px $green, inset 0 2px 1px rgba($black, .075);
    outline: none;
  }
}

.suggestionHighlighted {
  background: black;
  color: white;
  cursor: pointer;

  .suggestionInfo {
    color: white;
  }
}

.comboboxContainer {
  position: relative;
  width: 100%;

  * {
    box-sizing: border-box;
  }
}