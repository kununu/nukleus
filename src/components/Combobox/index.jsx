// NOTE: we make extensive use of property initialiser for binding methods to the class instance:
// onChange, getSuggestionValue etc.

import React from 'react';
import PropTypes from 'prop-types';
import Autosuggest from 'react-autosuggest';
import debounce from 'debounce';


import Error from '../Error';
import {
  controlLabel,
  controlNote,
  formControl,
  formControlError,
  formGroup,
  controlLabelRequired,
  hidden,
} from '../index.scss';

import styles from './index.scss';


export default class ComboboxComponent extends React.Component {
  static propTypes = {
    debounceRate: PropTypes.number,
    disabled: PropTypes.bool,
    error: PropTypes.string,
    errorSubInfo: PropTypes.string,
    handle: PropTypes.element,
    id: PropTypes.string.isRequired,
    inputProps: PropTypes.object, // eslint-disablie-line react/forbid-prop-types
    inputStyles: PropTypes.string,
    inputValue: PropTypes.string,
    isRequired: PropTypes.bool,
    isSearchable: PropTypes.bool,
    items: PropTypes.arrayOf(PropTypes.string),
    label: PropTypes.string,
    labelHidden: PropTypes.bool,
    name: PropTypes.string.isRequired,
    onBlur: PropTypes.func,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    onSelect: PropTypes.func,
    placeholder: PropTypes.string,
    requiredLabel: PropTypes.string,
  };

  static defaultProps = {
    debounceRate: 500,
    disabled: false,
    error: null,
    errorSubInfo: null,
    handle: null,
    inputProps: {},
    inputStyles: 'inline',
    inputValue: '',
    isRequired: false,
    isSearchable: true,
    items: [],
    label: '',
    labelHidden: false,
    onBlur: () => {},
    onChange: () => {},
    onFocus: () => {},
    onSelect: null,
    placeholder: '',
    requiredLabel: '',
  };

  state = {
    showError: false,
    suggestions: this.getSuggestions('', this.props.items), // eslint-disable-line react/destructuring-assignment
    value: this.props.inputValue, // eslint-disable-line react/destructuring-assignment
  };

  debouncedLoadSuggestions = debounce(this.loadSuggestions, this.props.debounceRate); // eslint-disable-line react/destructuring-assignment

  componentWillMount () {
    const {error} = this.props;

    // Show error, if already set
    if (error !== null) this.showError();
  }

  componentWillReceiveProps (nextProps) {
    const {inputValue} = this.props;

    if (nextProps.error) this.showError();
    if (nextProps.inputValue !== inputValue) {
      this.setState({value: nextProps.inputValue});
    }
  }

  onSuggestionsFetchRequested = ({value}) => {
    this.debouncedLoadSuggestions(value);
  };

  onChange = (event, {newValue}) => {
    const {
      isSearchable,
      items,
      onChange,
    } = this.props;

    onChange(event);
    if (!isSearchable && items.indexOf(newValue) === -1) {
      event.preventDefault();
      return;
    }
    this.hideError();
    this.setState({value: newValue});
  };

  getSuggestions (value, items = []) {
    const {isSearchable} = this.props;

    if (isSearchable) {
      const inputValue = value.trim().toLowerCase();
      const inputLength = inputValue.length;

      return items
        .filter(item => item && item.toLowerCase().slice(0, inputLength) === inputValue)
        .map(item => ({value: item}));
    }
    return items.map(item => ({value: item}));
  }

  getSuggestionValue = suggestion => suggestion.value;

  handleSelection = (e, {method, suggestionIndex, suggestionValue}) => {
    const {onSelect} = this.props;

    if (onSelect) onSelect(suggestionIndex, suggestionValue);
    if (method === 'enter') {
      e.preventDefault();
    }
  };

  loadSuggestions (value) {
    const {items} = this.props;

    this.setState({suggestions: this.getSuggestions(value, items)});
  }

  showError () {
    this.setState({showError: true});
  }

  hideError () {
    this.setState({showError: false});
  }

  hasError () {
    const {error} = this.props;
    const {showError} = this.state;

    return showError && error;
  }

  renderSuggestion = suggestion => <span>{suggestion.value}</span>;

  render () {
    const {
      name,
      id,
      label,
      labelHidden,
      error,
      errorSubInfo,
      handle,
      isRequired,
      requiredLabel,
      inputStyles,
      placeholder,
      disabled,
      isSearchable,
      inputProps,
      onBlur,
      onFocus,
    } = this.props;
    const {
      suggestions,
      value,
    } = this.state;

    return (
      <div
        className={`${formGroup} ${styles[inputStyles]} ${requiredLabel ? styles.paddingTop : ''}`}
        id={`${name}-container`}
      >
        {requiredLabel && (
        <span className={`${controlNote} ${controlLabelRequired}`}>
          {requiredLabel}
        </span>
        )}

        <label
          className={`${controlLabel} ${labelHidden && hidden} ${this.hasError() ? styles.controlLabelError : ''}`}
          htmlFor={id}
        >
          {label}
        </label>

        <div className={styles.container}>
          <Autosuggest
            suggestions={suggestions}
            theme={styles}
            error={error}
            errorSubInfo={errorSubInfo}
            onSuggestionSelected={this.handleSelection}
            onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
            shouldRenderSuggestions={() => true}
            onSuggestionsClearRequested={() => true}
            getSuggestionValue={this.getSuggestionValue}
            renderSuggestion={this.renderSuggestion}
            focusInputOnSuggestionClick={isSearchable}
            inputProps={{
              ...inputProps,
              className: `${formControl} ${!isSearchable && styles.isNotSearchable} ${this.hasError() ? formControlError : ''}`,
              disabled,
              id,
              name,
              onBlur,
              onChange: this.onChange,
              onFocus,
              placeholder,
              required: isRequired,
              value,
            }}
          />

          {handle ? (
            <span className={styles.handle}>
              {handle}
            </span>
          ) : ''}

          {this.hasError() && (
          <Error
            info={error}
            subInfo={errorSubInfo}
          />
          )}
        </div>
      </div>
    );
  }
}
