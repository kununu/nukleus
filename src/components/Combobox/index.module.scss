@import '~@kununu/kununu-theme-v2/scss/base/variables';

.comboboxContainer {
  position: relative;
  width: 100%;
  line-height: 1.65;

  * {
    box-sizing: border-box;
  }
}

.suggestionsContainer {
  box-shadow: 0 5px 10px rgba($black, .2);
  border-radius: $border-radius-base;
  background-clip: padding-box;
  font-size: $font-size-base;
  width: 100%;
  min-width: 250px;
  margin-top: 2px;
  background-color: $white;
  max-height: 300px;
  overflow: auto;
  position: absolute;
  top: 100%;
  left: 0;
  right: 0;
  z-index: 3;

  @media (max-width: $screen-xs-max) {
    left: 0;
    right: 0;
    border-radius: 0;
  }

  ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }
}

.suggestion {
  box-sizing: border-box;
  text-align: left;
  padding: 10px 20px;
}

.suggestionHighlighted {
  background: $blue;
  color: $white;
  cursor: pointer;
}

.comboboxHandle {
  position: absolute;
  right: 15px;
  font-size: 11px;
  color: $black;
  top: ($input-height-base / 3);
  z-index: 2;
  pointer-events: none;
}

.isNotSearchable {
  cursor: default;
  color : transparent;
  text-shadow : 0 0 0 $gray;
  user-select: none;
  &::selection { background: transparent; }

  &:focus {
    outline : none;
  }
}
