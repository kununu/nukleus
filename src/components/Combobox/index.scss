@import '~@kununu/kununu-theme-v2/scss/base/variables';

.combobobxContainer {
  position: relative;
  width: 100%;
  line-height: 1.65;

  * {
    box-sizing: border-box;
  }
}

.suggestionsContainer {
  box-shadow: 0 2px 10px rgba($black, .2);
  border-radius: $border-radius-base;
  background-clip: padding-box;
  font-size: $font-size-base;
  width: 100%;
  margin-top: 1px;
  background-color: $white;
  max-height: 300px;
  overflow: auto;
  position: absolute;
  top: 100%;
  left: 0;
  right: 0;
  z-index: 2;

  @media (max-width: $screen-xs-max) {
    width: auto;
    border-radius: 0;
  }

  ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }
}

.suggestion {
  box-sizing: border-box;
  text-align: left;
  padding: 8px 15px;
}

.suggestionHighlighted {
  background: $blue;
  color: $white;
  cursor: pointer;
}

.comboboxHandle {
  position: absolute;
  right: 15px;
  font-size: 11px;
  color: $black;
  top: ($input-height-base / 3);
  z-index: 2;
  pointer-events: none;
}

.isNotSearchable {
  cursor: default;
  color : transparent;
  text-shadow : 0 0 0 $gray;
  user-select: none;
  &::selection { background: transparent; }

  &:focus {
    outline : none;
  }
}
