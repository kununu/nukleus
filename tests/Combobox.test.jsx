import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import Combobox from 'Combobox'; // eslint-disable-line import/no-unresolved, import/extensions, import/no-extraneous-dependencies

const combobox = (
  <Combobox
    name="name"
    label="Combobox"
    id="name"
    isRequired
    placeholder="Type m"
    keyName="name"
    items={['music', 'maths', 'manga', 'morning', 'musical', 'mania', 'message', 'metal', 'micro', 'macro', 'microphone']} />
);

test('Renders Combobox without crashing', () => {
  const component = renderer.create(combobox);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Causes dropdown to show when input is focused', () => {
  const component = mount(combobox);
  component.find('input#name').simulate('focus');
  expect(toJson(component)).toMatchSnapshot();
});

const spyFunc = jest.fn();
const notSearchableCombobox = (
  <Combobox
    name="name"
    label="Combobox"
    id="name"
    isRequired
    placeholder="Type m"
    keyName="name"
    isSearchable={false}
    onSelect={spyFunc}
    items={['music', 'maths', 'manga', 'morning', 'musical', 'mania', 'message', 'metal', 'micro', 'macro', 'microphone']} />
);

test('Renders notSearchableCombobox without crashing', () => {
  const component = renderer.create(notSearchableCombobox);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Causes dropdown to show when input is focused', () => {
  const component = mount(notSearchableCombobox);
  component.find('input#name').simulate('focus');
  component.find('#react-autowhatever-1--item-0').simulate('click');
  expect(spyFunc).toHaveBeenCalled();
});
