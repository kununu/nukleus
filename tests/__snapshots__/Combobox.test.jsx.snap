exports[`test Causes dropdown to show when input is focused 1`] = `
<ComboboxComponent
  disabled={false}
  error={null}
  handle={null}
  id="name"
  inputStyles="inline"
  inputValue=""
  items={
    Array [
      "music",
      "maths",
      "manga",
      "morning",
      "musical",
      "mania",
      "message",
      "metal",
      "micro",
      "macro",
      "microphone",
    ]
  }
  keyName="name"
  label="Combobox"
  name="name"
  placeholder="Type m"
  required={true}>
  <div
    className="formGroup inline">
    <label
      className="controlLabel"
      htmlFor="city">
      Combobox
    </label>
    <div
      className="container">
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusFirstSuggestion={false}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        id="1"
        inputProps={
          Object {
            "className": "formControl",
            "disabled": false,
            "id": "name",
            "name": "name",
            "onChange": [Function],
            "placeholder": "Type m",
            "required": true,
            "value": "",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderSuggestion={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={
          Array [
            Object {
              "value": "music",
            },
            Object {
              "value": "maths",
            },
            Object {
              "value": "manga",
            },
            Object {
              "value": "morning",
            },
            Object {
              "value": "musical",
            },
            Object {
              "value": "mania",
            },
            Object {
              "value": "message",
            },
            Object {
              "value": "metal",
            },
            Object {
              "value": "micro",
            },
            Object {
              "value": "macro",
            },
            Object {
              "value": "microphone",
            },
          ]
        }
        theme={undefined {}}>
        <Autowhatever
          focusedItemIndex={null}
          focusedSectionIndex={null}
          getSectionItems={[Function]}
          id="1"
          inputProps={
            Object {
              "className": "formControl",
              "disabled": false,
              "id": "name",
              "name": "name",
              "onBlur": [Function],
              "onChange": [Function],
              "onFocus": [Function],
              "onKeyDown": [Function],
              "placeholder": "Type m",
              "required": true,
              "value": "",
            }
          }
          itemProps={[Function]}
          items={
            Array [
              Object {
                "value": "music",
              },
              Object {
                "value": "maths",
              },
              Object {
                "value": "manga",
              },
              Object {
                "value": "morning",
              },
              Object {
                "value": "musical",
              },
              Object {
                "value": "mania",
              },
              Object {
                "value": "message",
              },
              Object {
                "value": "metal",
              },
              Object {
                "value": "micro",
              },
              Object {
                "value": "macro",
              },
              Object {
                "value": "microphone",
              },
            ]
          }
          multiSection={false}
          renderInputComponent={[Function]}
          renderItem={[Function]}
          renderItemData={
            Object {
              "query": "",
            }
          }
          renderItemsContainer={[Function]}
          renderSectionTitle={[Function]}
          shouldRenderSection={[Function]}
          theme={Object {}}>
          <div
            style={Object {}}>
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-expanded={true}
              aria-haspopup={true}
              aria-owns="react-autowhatever-1"
              autoComplete="off"
              className="formControl"
              disabled={false}
              id="name"
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="Type m"
              required={true}
              role="combobox"
              style={Object {}}
              type="text"
              value="" />
            <div
              id="react-autowhatever-1"
              style={Object {}}>
              <ItemsList
                focusedItemIndex={null}
                getItemId={[Function]}
                itemProps={[Function]}
                items={
                  Array [
                    Object {
                      "value": "music",
                    },
                    Object {
                      "value": "maths",
                    },
                    Object {
                      "value": "manga",
                    },
                    Object {
                      "value": "morning",
                    },
                    Object {
                      "value": "musical",
                    },
                    Object {
                      "value": "mania",
                    },
                    Object {
                      "value": "message",
                    },
                    Object {
                      "value": "metal",
                    },
                    Object {
                      "value": "micro",
                    },
                    Object {
                      "value": "macro",
                    },
                    Object {
                      "value": "microphone",
                    },
                  ]
                }
                keyPrefix="react-autowhatever-1-"
                onFocusedItemChange={[Function]}
                renderItem={[Function]}
                renderItemData={
                  Object {
                    "query": "",
                  }
                }
                sectionIndex={null}
                theme={[Function]}>
                <ul
                  role="listbox"
                  style={Object {}}>
                  <Item
                    data-section-index={null}
                    data-suggestion-index={0}
                    id="react-autowhatever-1--item-0"
                    item={
                      Object {
                        "value": "music",
                      }
                    }
                    itemIndex={0}
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onTouchStart={[Function]}
                    renderItem={[Function]}
                    renderItemData={
                      Object {
                        "query": "",
                      }
                    }
                    sectionIndex={null}
                    style={Object {}}>
                    <li
                      data-section-index={null}
                      data-suggestion-index={0}
                      id="react-autowhatever-1--item-0"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onTouchStart={[Function]}
                      role="option"
                      style={Object {}}>
                      <span>
                        music
                      </span>
                    </li>
                  </Item>
                  <Item
                    data-section-index={null}
                    data-suggestion-index={1}
                    id="react-autowhatever-1--item-1"
                    item={
                      Object {
                        "value": "maths",
                      }
                    }
                    itemIndex={1}
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onTouchStart={[Function]}
                    renderItem={[Function]}
                    renderItemData={
                      Object {
                        "query": "",
                      }
                    }
                    sectionIndex={null}
                    style={Object {}}>
                    <li
                      data-section-index={null}
                      data-suggestion-index={1}
                      id="react-autowhatever-1--item-1"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onTouchStart={[Function]}
                      role="option"
                      style={Object {}}>
                      <span>
                        maths
                      </span>
                    </li>
                  </Item>
                  <Item
                    data-section-index={null}
                    data-suggestion-index={2}
                    id="react-autowhatever-1--item-2"
                    item={
                      Object {
                        "value": "manga",
                      }
                    }
                    itemIndex={2}
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onTouchStart={[Function]}
                    renderItem={[Function]}
                    renderItemData={
                      Object {
                        "query": "",
                      }
                    }
                    sectionIndex={null}
                    style={Object {}}>
                    <li
                      data-section-index={null}
                      data-suggestion-index={2}
                      id="react-autowhatever-1--item-2"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onTouchStart={[Function]}
                      role="option"
                      style={Object {}}>
                      <span>
                        manga
                      </span>
                    </li>
                  </Item>
                  <Item
                    data-section-index={null}
                    data-suggestion-index={3}
                    id="react-autowhatever-1--item-3"
                    item={
                      Object {
                        "value": "morning",
                      }
                    }
                    itemIndex={3}
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onTouchStart={[Function]}
                    renderItem={[Function]}
                    renderItemData={
                      Object {
                        "query": "",
                      }
                    }
                    sectionIndex={null}
                    style={Object {}}>
                    <li
                      data-section-index={null}
                      data-suggestion-index={3}
                      id="react-autowhatever-1--item-3"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onTouchStart={[Function]}
                      role="option"
                      style={Object {}}>
                      <span>
                        morning
                      </span>
                    </li>
                  </Item>
                  <Item
                    data-section-index={null}
                    data-suggestion-index={4}
                    id="react-autowhatever-1--item-4"
                    item={
                      Object {
                        "value": "musical",
                      }
                    }
                    itemIndex={4}
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onTouchStart={[Function]}
                    renderItem={[Function]}
                    renderItemData={
                      Object {
                        "query": "",
                      }
                    }
                    sectionIndex={null}
                    style={Object {}}>
                    <li
                      data-section-index={null}
                      data-suggestion-index={4}
                      id="react-autowhatever-1--item-4"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onTouchStart={[Function]}
                      role="option"
                      style={Object {}}>
                      <span>
                        musical
                      </span>
                    </li>
                  </Item>
                  <Item
                    data-section-index={null}
                    data-suggestion-index={5}
                    id="react-autowhatever-1--item-5"
                    item={
                      Object {
                        "value": "mania",
                      }
                    }
                    itemIndex={5}
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onTouchStart={[Function]}
                    renderItem={[Function]}
                    renderItemData={
                      Object {
                        "query": "",
                      }
                    }
                    sectionIndex={null}
                    style={Object {}}>
                    <li
                      data-section-index={null}
                      data-suggestion-index={5}
                      id="react-autowhatever-1--item-5"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onTouchStart={[Function]}
                      role="option"
                      style={Object {}}>
                      <span>
                        mania
                      </span>
                    </li>
                  </Item>
                  <Item
                    data-section-index={null}
                    data-suggestion-index={6}
                    id="react-autowhatever-1--item-6"
                    item={
                      Object {
                        "value": "message",
                      }
                    }
                    itemIndex={6}
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onTouchStart={[Function]}
                    renderItem={[Function]}
                    renderItemData={
                      Object {
                        "query": "",
                      }
                    }
                    sectionIndex={null}
                    style={Object {}}>
                    <li
                      data-section-index={null}
                      data-suggestion-index={6}
                      id="react-autowhatever-1--item-6"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onTouchStart={[Function]}
                      role="option"
                      style={Object {}}>
                      <span>
                        message
                      </span>
                    </li>
                  </Item>
                  <Item
                    data-section-index={null}
                    data-suggestion-index={7}
                    id="react-autowhatever-1--item-7"
                    item={
                      Object {
                        "value": "metal",
                      }
                    }
                    itemIndex={7}
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onTouchStart={[Function]}
                    renderItem={[Function]}
                    renderItemData={
                      Object {
                        "query": "",
                      }
                    }
                    sectionIndex={null}
                    style={Object {}}>
                    <li
                      data-section-index={null}
                      data-suggestion-index={7}
                      id="react-autowhatever-1--item-7"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onTouchStart={[Function]}
                      role="option"
                      style={Object {}}>
                      <span>
                        metal
                      </span>
                    </li>
                  </Item>
                  <Item
                    data-section-index={null}
                    data-suggestion-index={8}
                    id="react-autowhatever-1--item-8"
                    item={
                      Object {
                        "value": "micro",
                      }
                    }
                    itemIndex={8}
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onTouchStart={[Function]}
                    renderItem={[Function]}
                    renderItemData={
                      Object {
                        "query": "",
                      }
                    }
                    sectionIndex={null}
                    style={Object {}}>
                    <li
                      data-section-index={null}
                      data-suggestion-index={8}
                      id="react-autowhatever-1--item-8"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onTouchStart={[Function]}
                      role="option"
                      style={Object {}}>
                      <span>
                        micro
                      </span>
                    </li>
                  </Item>
                  <Item
                    data-section-index={null}
                    data-suggestion-index={9}
                    id="react-autowhatever-1--item-9"
                    item={
                      Object {
                        "value": "macro",
                      }
                    }
                    itemIndex={9}
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onTouchStart={[Function]}
                    renderItem={[Function]}
                    renderItemData={
                      Object {
                        "query": "",
                      }
                    }
                    sectionIndex={null}
                    style={Object {}}>
                    <li
                      data-section-index={null}
                      data-suggestion-index={9}
                      id="react-autowhatever-1--item-9"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onTouchStart={[Function]}
                      role="option"
                      style={Object {}}>
                      <span>
                        macro
                      </span>
                    </li>
                  </Item>
                  <Item
                    data-section-index={null}
                    data-suggestion-index={10}
                    id="react-autowhatever-1--item-10"
                    item={
                      Object {
                        "value": "microphone",
                      }
                    }
                    itemIndex={10}
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onTouchStart={[Function]}
                    renderItem={[Function]}
                    renderItemData={
                      Object {
                        "query": "",
                      }
                    }
                    sectionIndex={null}
                    style={Object {}}>
                    <li
                      data-section-index={null}
                      data-suggestion-index={10}
                      id="react-autowhatever-1--item-10"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onTouchStart={[Function]}
                      role="option"
                      style={Object {}}>
                      <span>
                        microphone
                      </span>
                    </li>
                  </Item>
                </ul>
              </ItemsList>
            </div>
          </div>
        </Autowhatever>
      </Autosuggest>
    </div>
  </div>
</ComboboxComponent>
`;

exports[`test Renders Combobox without crashing 1`] = `
<div
  className="formGroup inline">
  <label
    className="controlLabel"
    htmlFor="city">
    Combobox
  </label>
  <div
    className="container">
    <div
      style={Object {}}>
      <input
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-expanded={false}
        aria-haspopup={false}
        aria-owns="react-autowhatever-1"
        autoComplete="off"
        className="formControl"
        disabled={false}
        id="name"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder="Type m"
        required={true}
        role="combobox"
        style={Object {}}
        type="text"
        value="" />
      <div
        id="react-autowhatever-1"
        style={Object {}} />
    </div>
    
  </div>
</div>
`;
